Lindblad Master Equation

PhilipVinc opened the pull request:
~~This is still WIP, but I would like some preliminary feedback on this as it requires some changes to core netket classes.~~

The PR mainly updates the density matrix to the new interface and gets rid of lookup tables, and defines a new operator, `LocalLindblad` which will behave like a standard operator with a doubled space.

My aim at the moment is to get the Lindblad ME to work with what is currently implemented at the moment. This mainly means _faking_ all methods such as `LogVal`, `Nvisible`... to behave as if it was a standard machine defined on a doubled space.

Connections (in operators) are handled similarly, by assuming that rows are the first half entries of the configuration and columns are the second half. 

Changes to netket core:
 - `AbstractHilbert` has a new virtual method called `UpdateConfRowCol` to efficiently update the row and columns of a configuration given the `connections, to change` from a doubled Hilbert space. There is a standard (slow) implementation, and more efficient ones are defined for `Spins` and `CustomHilbert`.
- Add `LocalOperator` copy constructor, which was previously commented out. I need it to initialize the Lindblad.

as of now, in python it is possible to construct a machine and do the following:
```python
import netket as nk
import numpy as np
from netket.machine import NdmSpinPhase
from netket.hilbert import Spin
from netket.graph import Hypercube

vv = np.random.rand(18)
vl = vv[0:9]
vr = vv[9:18]

vv2 = np.random.rand(3, 18)
vl2 = vv2[:,0:9]
vr2 = vv2[:,9:18]

g = Hypercube(length=9, n_dim=1)
hi = Spin(s=0.5, total_sz=0.5, graph=g)
ma = NdmSpinPhase(hilbert=hi,alpha=1, beta=2)
ma.init_random_parameters()
assert (ma.log_val(vv) == ma.log_val(vl, vr))
assert (ma.log_val(vv2) == ma.log_val(vl2, vr2)).all()

assert (ma.der_log(vv) == ma.der_log(vl, vr)).all()
assert (ma.der_log(vv2) == ma.der_log(vl2, vr2)).all()
```
Right now you can also pass the density matrix to a sampler and it will sweep it correctly.

construct the machine sampling the diagonal of the density matrix
```python
dd = ma.diagonal()
dd.log_val(vl)
```

For how to construct the Liouvillian, at the moment you pass the hamiltonian and then add every jump operator one by one. I'd like to make him accept a list of operators, but I did not yet figure out how to do the python bindings for this one. Who's the expert on python bindings?


```python
ha = nk.operator.LocalOperator(hi)

for i in range(L):
    ha += nk.operator.LocalOperator(hi, sx, [i])
    ha += nk.operator.LocalOperator(hi, np.kron(sz, sz), [i, (i + 1) % L])


# Create the lindbladian with no jump operators
lind = nk.operator.LocalLindbladian(ha)

# Add a sigmam jump operator on each site
for i in range(L):
    j_op = nk.operator.LocalOperator(hi, sigmam, [i])
    lind.add_jump_op(j_op)
```

The method `FindConn` of `LocalLindbladian` is an horrible mess that I have to figure out how to simplify...

Lastly: the VMC driver. I either plug into the existing driver and plug a lot of if/else statements checking for the class, or I create another driver (my preferred take).
If I create another driver, there will be a lot of shared functionality among the two, so it would make sense to create a base python class from which the two derive, but if you are already working on that part of netket It's best too coordinate a bit.

F

gcarleo commented:
Thanks a lot @PhilipVinc ! Just a preliminary comment: yes, for the driver we will need to write (a separate one) in pure python, as much as done with the latest vmc driver. It is true that there is some code overlap between drivers, but we haven't thought too much about how to remove that yet. Actually, the key parts to be reused among drivers concern the handling of the output, and that is yet to be implemented beyond the legacy json format. So, for the moment I'd suggest to write a minimal lindblad driver separately (basically without handling the output part), and later think about sharing code with the other drivers.

gcarleo commented:
Just another thought (even though I am sure this is already taken care of): before writing the stochastic optimization driver, can we test the Lindbladian using ED ? I guess that there should already be a method to_sparse (or to_dense) for the Lindbladian, so that might give us a good handle to debug possible issues with the find_conn

PhilipVinc commented:
@gcarleo Yep. The methods `to_sparse` and `to_dense` return the matrix representation of the linouvillian superop. They work out of the box as `Lindblad` is also an operator. For reference, the ordering of tensor products is based upon [arXiv:1504.05266](https://arxiv.org/pdf/1504.05266.pdf) and is the same ordering used by qutip. 

I just implemented a method named `nk.exact.steadystate_ed` which takes as input the lindbladian and diagonalizes `LdagL`.
ED only works for 4/5 spins, though. It would be nice to add another exact method based on bicg_stab solving L\rho = 0, which can go up to 13/14 sites, but I'll leave that for later

PhilipVinc commented:
I have also added a new constructor to `CustomGraph`, taking only the number of sites as input. 
This is needed to construct the Doubled Hilbert space for systems with only one site, where there are no edges.

gcarleo commented:
> I have also added a new constructor to CustomGraph, taking only the number of sites as input.
> This is needed to construct the Doubled Hilbert space for systems with only one site, where there are no edges.

It might be helpful also to know that there is a `EdgelessGraph` now, that is constructed only with the number of sites

PhilipVinc commented:
@gcarleo Is the functionality to compute the derivative of the local energy terms (such as that you need to perform variance minimisation) exposed in python?

femtobit commented:
> Lastly: the VMC driver. I either plug into the existing driver and plug a lot of if/else statements checking for the class, or I create another driver (my preferred take).
If I create another driver, there will be a lot of shared functionality among the two, so it would make sense to create a base python class from which the two derive, but if you are already working on that part of netket It's best too coordinate a bit.

Looking at the current (after @gcarleo's PR introducing the Python sampler interface) driver code https://github.com/netket/netket/blob/543b07f42181b25395ff55086903ca72f852b91c/netket/_vmc.py#L83-L313
I think you would [rimarily need to adapt `__init__` and `advance`, right? Are there other changes you would need compared to the current driver?

`Vmc.advance` now contains mostly code related to sampling, which is probably very similar for your case. So maybe that part should be extracted too.

(Moving forward it still might be easiest if you write another driver based on `Vmc` and, once it works, we can extract the common functionality.)

PhilipVinc commented:
> It might be helpful also to know that there is a `EdgelessGraph` now, that is constructed only with the number of sites

I'd need to add a custom constructor to `EdgelessGraph` so that the automorphisms are only built within the row and column subspaces. 
Is this ok?

gcarleo commented:
> @gcarleo Is the functionality to compute the derivative of the local energy terms (such as that you need to perform variance minimisation) exposed in python?

No, not really, sorry.... maybe we can discuss about doing this efficiently here or in another thread

gcarleo commented:
> I'd need to add a custom constructor to `EdgelessGraph` so that the automorphisms are only built within the row and column subspaces.
> Is this ok?

But `EdgelessGraph` is not supposed to have automorphism I guess...

PhilipVinc commented:
> No, not really, sorry.... maybe we can discuss about doing this efficiently here or in another thread

Then let's do it. Without it I can't really move forward.

gcarleo commented on Examples/ExactDiag/dissipative_ising1.py:
```diff
@@ -0,0 +1,55 @@
+    lind.add_jump_op(j_op)
+
+
+rho = nk.exact.steadystate_ed(lind)
```
not sure you need to specify `ed` since this is already in the `exact` module, I guess taht 
`nk.exact.steady_state` would work?

PhilipVinc replied:
I'd like to eventually implement two methods, `ED` and `iterative`.

gcarleo replied:
I'd put that as an option (i.e. `nk.exact.steady_state(..,method='iterative')`), but ok, that's just a minor naming issue we can discuss later

PhilipVinc replied:
seems good

PhilipVinc commented:
@femtobit Yeah, for the optimisation I will mostly need to change the sampling part (It's more similar to a variance minimization). For the observables I also need to keep a separate sampler/chains/states, as they are sampled differently.

Can I ask you to give me an hand [on this](https://github.com/netket/netket/pull/315/files#diff-2319c9bfd22345de68b4a846ca88286bR62) line? 
I basically want to insert a `nonstd::span` at the end of a vector. Is it even possible?

femtobit commented:
> Can I ask you to give me an hand on this line?

`connectors` is a `vector<vector<int>>` (so each element is a `vector<...>` owning a part of memory). On the other hand, `conn.tochange` is a `span` and thus references memory which it does not own. So to append the data to `connectors`, you need to create a new vector (which you append to `connectors`) and copy the data there.

Originally, I added `ForEachConn` to get rid of all the `vector` of `vector` structures. Couldn't you implement `ForEachConn` for you Lindbladian? That would give you `FindConn`, as it can be easily implemented in terms of `ForEachConn`

Otherwise, there is no way to get rid of allocating new vectors and appending them.

femtobit commented:
@PhilipVinc I think for your Lindbladian, `ForEachConn` should look approximately like this. Note how in many cases, the `span`s from the underlying operator's `ForEachConn` can be passed through. For operations that require changing the values (i.e., the addition of `N` for the columns), there is still a copy, but this version only needs one `std::vector`.
```cpp
void LocalLindbladian::ForEachConn(VectorConstRefType v, ConnCallback callback) const {
  const Complex im{0, 1};
  const Index physical_size = GetHilbertDoubled().SizePhysical();
  auto vrow = v.head(physical_size);
  auto vcol = v.tail(physical_size);

  Hnh_dag_.ForEachConn(vrow, [&](ConnectorRef conn) {
    callback({-im * conn.mel, conn.tochange, conn.newconf});
  });

  // Needed to store modified configurations in some places
  std::vector<int> tochange;

  Hnh_.ForEachConn(vcol, [&](ConnectorRef conn) {
    tochange.clear();
    std::transform(conn.tochange.begin(),
                   conn.tochange.end(),
                   std::back_inserter(tochange),
                   [physical_size](int idx) {
                     return idx + physical_size;
                   });
    callback({-im * conn.mel, tochange, conn.newconf});
  });

  // Compute term \sum_i L_i \otimes L_i^\dagger
  // Iterate over all jump operators
  for (size_t i = 0; i < jump_ops_.size(); i++) {
    const auto& op = jump_ops_[i];

    // For every connection <vrow|L_i|x>
    op.ForEachConn(vrow, [&](ConnectorRef conn_row) {
      // For every connection <vrow|L_i^\dagger|x>
      tochange.clear();
      op.ForEachConn(vcol, [&](ConnectorRef conn_col) {
        std::transform(conn_col.tochange.begin(),
                   conn_col.tochange.end(),
                   std::back_inserter(tochange),
                   [physical_size](int idx) {
                     return idx + physical_size;
                   });

        // FIXME:
        /*callback({std::conj(conn_row.mel) * conn_col.mel, ..., ...});*/
      });
    });
  }
}
```

PhilipVinc commented:
@femtobit Thanks! It seems I can't avoid creating a copy here.

I went for something similar: I created a `ForEachConnSuperOp` to work at the super-operatorial level (where I don't need copies), and then for the operatorial-level stuff such as `ForEachConn` I do create only two temporaries, holding the concatenated `to change` and `newconf`.

For now the superoperatorial stuff is not used, but eventually I could add some specialised more efficient methods.

gcarleo commented on Sources/Operator/py_operator.hpp:
```diff
@@ -172,6 +173,56 @@ void AddOperatorModule(py::module m) {
+          case 2:
+            return py::cast(DerLocalValues(
+                Eigen::Map<const RowMatrix<double>>{
+                    samples.data(), samples.shape(0), samples.shape(1)},
```
shouldn't there be the number of parameters in the shape somewhere?

PhilipVinc replied:
yes, thanks

gcarleo commented on Sources/Operator/abstract_operator.cc:
```diff
@@ -86,4 +86,38 @@ Eigen::VectorXcd LocalValues(Eigen::Ref<const RowMatrix<double>> samples,
+    RowMatrix<Complex> der_log_diff = machine.DerLogDiff(v, tochange, newconfs);
+
+    Eigen::Map<const Eigen::ArrayXcd> meleig(&mel[0], mel.size());
+    Eigen::VectorXcd loc = meleig * outlvd.array().exp();
```
maybe these temporary loc and aa objects should be declared out of the loop, assuming the compiler does not optimize them out

PhilipVinc commented:
@gcarleo How do I create from python a sampler with a certain `machine_func`? When I sample the diagonal of the density matrix, for observables,  I need `machine_func` to be the identity (or `real`).

gcarleo commented:
> @gcarleo How do I create from python a sampler with a certain machine_func? When I sample the diagonal of the density matrix, for observables, I need machine_func to be the identity (or real).

in python you just do 
```python
my_sampler.machine_func=lambda x : np.abs(x) 
```
or something like that...

in c++ samplers have the method set_machine_func that accepts a generic std::function

PhilipVinc commented:
thanks but
```
>>> sa_obs.machine_func=lambda x : np.abs(x)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: <lambda>() takes 1 positional argument but 2 were given
```

gcarleo commented:
ah sorry, it's a function taking in,out arguments, numpy style. Just do 

```python
lambda x,out : np.abs(x,out)
```

gcarleo commented:
@PhilipVinc I have merged the changes on samplers, please try to pull the new stuff and see if you don't have conflicts (you shouldn't, but just in case).

gcarleo commented on netket/_ss.py:
```diff
@@ -0,0 +1,411 @@
+        return str(obj)
+
+
+def make_optimizer_fn(arg, ma):
```
we should move this to the optimizer module and also remove it from _vmc...

femtobit replied:
I think the best option would be to have a submodule with shared functions (and possibly a base class) for the driver classes. The `make_..._fn` are specifically written for the needs of the driver.

femtobit replied:
Also, can we give `_ss.py` a more descriptive name? Abbreviating `_vmc.py` is fine, since VMC is usually used as an acronym. I don't think that's treu for *steady state*. Why not call it `_steadystate.py`.

gcarleo replied:
> I think the best option would be to have a submodule with shared functions (and possibly a base class) for the driver classes. The `make_..._fn` are specifically written for the needs of the driver.

I agree, but maybe it's best to do this in another PR. Actually I think we should just remove make_... and instead have converrters from (say) optimizer x to netket optimizer

PhilipVinc commented:
I slightly changed the implementation to something more performant.

I had to add to `nk.stats` a method to compute the mean (and average it with MPI).

For the rest... I also added an example to Examples/DissipativeIsing1d. Seems that everything is working.

I'll go back and check if some docstrings are missing when I have time, but otherwise this PR is finished for me.

(In another PR we should split the optimiser logic...)

gcarleo commented on netket/stats.py:
```diff
@@ -8,3 +8,11 @@ def subtract_mean(x):
+
+def compute_mean(x):
+    """
+    Subtracts the mean of the input array over all but the last dimension
```
please update the docstring

gcarleo commented on Sources/Hilbert/spins.cc:
```diff
@@ -142,4 +142,23 @@ void Spin::UpdateConf(Eigen::Ref<Eigen::VectorXd> v,
   }
 }
 
+void Spin::UpdateConfRowCol(Eigen::Ref<Eigen::VectorXd> vrow,
```
What is the purpose of this function? sorry I got lost...

PhilipVinc replied:
it's equivalent to UpdateConf, but takes two visible configurations, `vrow` and `vrow`, and when `tochange[i] < NVisible()` it updates `vrow` otherwise it updates `vcol`.

PhilipVinc replied:
Resolved by removing this method

PhilipVinc commented on Sources/Operator/local_liouvillian.hpp:
```diff
@@ -0,0 +1,143 @@
+   * @param jump_ops : a vector of LocalOperators.
+   */
+  explicit LocalLiouvillian(const LocalOperator &H,
+      const std::vector<const LocalOperator> &jump_ops);
```
@femtobit Could you point me in the right direction in how to convert a python list (containing only LocalOperators) to a vector of const in order to bind this constructor to python?

PhilipVinc replied:
NVM, It's just a compilation problem. I figured it out..

gcarleo commented on setup.py:
```diff
@@ -1,5 +1,6 @@
 import os
 import platform
+import multiprocessing
```
not sure this speeds up compiling, it's mostly cmake that runs in parallel already?

PhilipVinc replied:
My bad. I did not want to commit that (It's supposed to be an RFC on another PR). I'll revert it.

it does speed things up things significantly (for me at least, on macOs). when you call cmake build it does not run in parallel by default. The problem is that this command is only available in a slightly more recent CMake version than the one we currently require.

gcarleo replied:
OK cool, then we should upgrade the required CMake version

PhilipVinc replied:
Currently we require cmake `3.10.3`. `--parallel` requires `3.12`. Current version is `3.16`.
According to cmake's website `3.10.3` was released on March 18, 2018,  while `3.12` is dated July 17, 2018. So it's just 4 months more recent.

PhilipVinc replied:
I'll bump the required version on `CMakeLists.txt` then?

gcarleo commented on Examples/DissipativeIsing1d/ising1d.py:
```diff
@@ -0,0 +1,84 @@
+op = nk.optimizer.Sgd(0.01)
+sr = nk.optimizer.SR(diag_shift=0.01, use_iterative=True)
+
+ss = nk.Steadystate(lind, sa, op, 2000, sampler_obs=sa_obs, n_samples_obs=500)
```
small thing, but this should be SteadyState, to be consistent with the camel case rules we have used elsewhere

gcarleo commented on Examples/ExactDiag/dissipative_ising1.py:
```diff
@@ -0,0 +1,55 @@
+    lind.add_jump_op(j_op)
+
+
+rho = nk.exact.steadystate(lind)
```
sorry again for these remarks, but they are kind of important for consistency: this should be exact.steady_state

gcarleo commented on netket/exact.py:
```diff
@@ -171,6 +171,45 @@ def full_ed(operator, first_n=1, compute_eigenvectors=False):
+            from scipy.sparse.linalg import eigsh
+
+            lind_mat = lindblad.to_sparse()
+            ldagl = lind_mat.H * lind_mat
```
isn't there a way to do ED without forming this dense matrix?

PhilipVinc replied:
You can compute the spectrum of `L`, but it's non hermitian so it often comes with big errors.
The alternative would be an iterative method solving L\rho=0, which I plan to implement at some point

gcarleo commented on Sources/Operator/local_operator.hpp:
```diff
@@ -201,27 +203,39 @@ class LocalOperator : public AbstractOperator {
+  void FindConn(VectorConstRefType v, std::vector<Complex> &mel,
+                std::vector<std::vector<int>> &connectors,
+                std::vector<std::vector<double>> &newconfs,
+                bool clear_new) const {
```
is clear_new used somewhere?

PhilipVinc replied:
Good catch. It's a leftover from a previous implementation of LocalLiouvillian. It's unneeded so I'll remove it.

gcarleo commented on Sources/Operator/py_operator.hpp:
```diff
@@ -170,8 +173,109 @@ void AddOperatorModule(py::module m) {
+      py::arg{"batch_size"} = 16,
+      R"EOF(Computes local values of the operator `op` for all `samples`.
+
+            Args:
```
please update this comment

gcarleo commented:
Whoops, sorry @PhilipVinc , there is a few conflicts, let me know if you need help with this
I promise we can merge this after, sorry

gcarleo commented on netket/operator.py:
```diff
@@ -155,12 +153,142 @@ def local_values(op, machine, v, log_vals=None, out=None):
+
+def der_local_values(op, machine, v, log_vals=None, der_log_vals=None, out=None, center_derivative=True):
+    """
+    Computes local values of the operator `op` for all `samples`.
```
please update docs

gcarleo commented on Sources/Stats/mc_stats.cc:
```diff
@@ -173,10 +173,15 @@ Eigen::VectorXcd product_sv(Eigen::Ref<const Eigen::VectorXcd> s_values,
 }
 
-void SubtractMean(Eigen::Ref<RowMatrix<Complex>> v_values) {
+Eigen::VectorXcd ComputeMean(Eigen::Ref<RowMatrix<Complex>> v_values) {
```
we don't need these functions anymore, right?

PhilipVinc replied:
they are still called from python in `stats.subtract_mean`. Can easily be deleted though

PhilipVinc replied:
I got rid of those.

gcarleo commented on setup.py:
```diff
@@ -165,5 +167,5 @@ def _decode(x):
     cmdclass=dict(build_ext=CMakeBuild),
     zip_safe=False,
-    install_requires=["numpy>=1.16", "cmake>=3.10.3", "scipy>=1.2.1", "mpi4py>=3.0.1"],
+    install_requires=["numpy>=1.16", "cmake>=3.12", "scipy>=1.2.1", "mpi4py>=3.0.1"],
```
too new cmake?

PhilipVinc replied:
We had this discussion at some point.
I enabled in `setup.py` the `--parallel` flag when you call `cmake build`. But the flag is only available in cmake>=3.12 (which is ~ 9 months more recent than 3.10.3).

`--parallel` is somewhat equivalent to the `-- -j 4` that you recently merged, but has the additional nice thing that it picks automatically the number of processors and if some build system does not like -j 4 he adapts to it.

I can revert this, though, if you prefer.

PhilipVinc replied:
So you want this to stay, or you want this to go?

gcarleo replied:
I think we can leave it to 3.12 !

PhilipVinc commented:
@gcarleo I corrected the docstrings that you had mentioned, and disabled the C++ kernel that was responsible for the bad performance. I also added complete tests to the driver.

While I still see a performance degradation of `der_local_values` compared to my original pure-C++ implementation (~25-30%), and there's still space for improvements, thanks to various improvements since your last PR, the total runtime is only 16% worse than before your PR.

I'd be glad if we merged this, and leave performance improvements for the future.

gcarleo commented:
OK thanks a lot @PhilipVinc , I think it's time to merge, I agree... Indeed we can later focus on rewriting these kernels in Numba and/or doing other tricks with GetConn. 

Despite the performance hit (that's still not too bad at this stage), we have gained a lot in generality and flexibility, and that is something really valuable!
